publisher/lock/s3_lock.go

```diff
@@ -55,37 +54,60 @@ func (l *lockData) isExpired(ttl time.Duration, t time.Time) bool {
 	return l.CreatedAt.Add(ttl).Before(t)
 }
 
+func NewS3Config(bucketName, roleARN, awsRegion, tags, lockGroup, owner string, maxRetries uint, retryBackoff, ttl time.Duration) S3Config {
+	return S3Config{
+		Bucket:       bucketName,
+		RoleARN:      roleARN,
+		Region:       awsRegion,
+		Tags:         tags,
+		Filepath:     lockGroup,
+		Owner:        owner,
+		TTL:          ttl,
+		RetryBackoff: retryBackoff,
+		MaxRetries:   maxRetries,
+	}
+}
+
 // NewS3 creates a lock instance ready to be used validating required AWS credentials.
-func NewS3(bucket, roleARN, region, filepath, owner string) (*S3, error) {
+func NewS3(c S3Config, logfn Logf) (*S3, error) {
 	sess, err := session.NewSession()
 	if err != nil {
 		return nil, err
 	}
 
-	creds := stscreds.NewCredentials(sess, roleARN, func(p *stscreds.AssumeRoleProvider) {})
-	conf := aws.Config{
+	creds := stscreds.NewCredentials(sess, c.RoleARN, func(p *stscreds.AssumeRoleProvider) {})
+	awsCfg := aws.Config{
 		Credentials: creds,
-		Region:      aws.String(region),
+		Region:      aws.String(c.Region),
 	}
 
 	return &S3{
-		client:   s3.New(sess, &conf),
-		owner:    owner,
-		bucket:   bucket,
-		filePath: filepath,
-		tags:     fmt.Sprintf("department=product&product=%s&project=%s&owning_team=%s&environment=%s", tagProduct, tagProject, tagOwningTeam, tagEnv),
-		ttl:      defaultTTL,
+		client: s3.New(sess, &awsCfg),
+		logF:   logfn,
+		conf:   c,
 	}, nil
 }
 
 // Lock S3 has no compare-and-swap so this is no bulletproof solution, but should be good enough.
 func (l *S3) Lock() error {
+	var tries uint
+retry:
+	if (l.conf.MaxRetries - tries) > 0 {
+		tries++
+		l.logF("%s attempt %d", l.conf.Owner, tries)
+		if l.isBusyDeletingExpired() {
+			l.logF("%s failed, waiting %s", l.conf.Owner, l.conf.RetryBackoff.String())
+			time.Sleep(l.conf.RetryBackoff)
+			goto retry
```

Comment:
I would replace the `goto` for a `for`+break loop iterating over the MaxRetries just to be a more easy to read. btw i didn't know that `goto` exists on Go. Just a suggestion.

Reply:
https://github.com/newrelic/infrastructure-publish-action/pull/85